### ISSUE TITLE
Docker login fails with "warning msg="error authorizing context: authorization token required" "

### ISSUE THREAD

**Issue by yannhowe** [bitnami/bitnami-docker-harbor-portal#8]
**Description**

Unable to docker login

**Steps to reproduce the issue:**

```
# follow instructions
curl -LO https://raw.githubusercontent.com/bitnami/bitnami-docker-harbor-portal/master/docker-compose.yml
curl -L https://github.com/bitnami/bitnami-docker-harbor-portal/archive/master.tar.gz | tar xz --strip=1 --wildcards '*-master/config'

# Generate self signed certs
sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout harbor/reg.mydomain.com.key -out harbor/reg.mydomain.com.crt
sudo cp reg.mydomain.com.crt /usr/local/share/ca-certificates/
sudo update-ca-certificates --fresh

```

**Changes to default docker-compose and configs**
`registry/config.yml`
```
auth:
  token:
    issuer: harbor-token-issuer
    realm: https://reg.mydomain.com/service/token # updated to HTTPS
    rootcertbundle: /etc/registry/root.crt
    service: harbor-registry
```

`docker-compose.yml`
```
services:
  core:
    environment:
      - EXT_ENDPOINT=https://reg.mydomain.com # Updated to HTTPS
  harbor-nginx:
    volumes:
      - ./data/ssl:/etc/ssl/ # Added to include SSL certs
    ports:
      - '443:8080' # changed to 443
```

`proxy/nginx.conf`
```
  server {
    listen 8080 ssl; # added ssl
    ssl_certificate     /etc/ssl/reg.mydomain.com.crt; # added
    ssl_certificate_key /etc/ssl/reg.mydomain.com.key; # added
  }
```

`curl -i -k -u admin:bitnami https://reg.mydoman.com/service/token\?service\=harbor-registry`
```
@ubuntu:~$ curl -i -k -u admin:bitnami https://reg.mydomain.com/service/token\?service\=harbor-registry
HTTP/1.1 200 OK
Server: nginx
Date: Sun, 12 Jul 2020 12:26:11 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 1105
Connection: keep-alive
Set-Cookie: sid=6ee6c7a56b46c364ce551e468fe694e8; Path=/; HttpOnly
X-Request-Id: 76e648eb-d89c-49be-8d95-2ed6d6c4bfe6

{"token":"eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjczSU46NEhSNjpaQkVYOjZOWE46MjNGSDoyWktEOkdQRFM6S1RPSzpZNVZPOjdJSFM6UFJWSzpIR0dHIn0.eyJpc3MiOiJoYXJib3ItdG9rZW4taXNzdWVyIiwic3ViIjoiYWRtaW4iLCJhdWQiOiJoYXJib3ItcmVnaXN0cnkiLCJleHAiOjE1OTQ1NTg1NzEsIm5iZiI6MTU5NDU1Njc3MSwiaWF0IjoxNTk0NTU2NzcxLCJqdGkiOiJyZXVjVkpsNGE5NHdUeFZyIiwiYWNjZXNzIjpudWxsfQ.SbO8xdHMjmRsX58fUS92oesaS36A7DRak_MMT8XNPE5GKjfrK25kFVSYbY3nZs8e3t56h0kFAYkPJ9_CJI3biPt4qWvyUHs7nNbjBxhU6iD1DU-MVKM5siK6OheIhviPF_Ger6pyYrL30Zs5DRGT1SdDvKmSm_NOqGspxejBAXwZKxpnkn1wEnbp4h4Vam2naCqPWmA34gZ8KplYY6pHvKCaoFiYWO9_hoKsBgKa1iserMY14sFFDcSc8rJISNbZN4u5X1pYh4QARqtXR6NXAjfwVqHmQvTBEucumfr4zFvFv_4yYrGY4yYkZmRalzw4DVtWQSIk6mY3gnTtth35qSNlfk5NtyHTOc_daNFTF2d47YBCu7WDHVmKZ5Tm5oJDHSJgxexhfCMPEccMpckXDctlseiqYHh--ZpYtVOMAPbHlcbZV8sy03ciwhdubZvU6vcBlA0eUk418l_7Okcmu7olYBZi2OhVjUPzoO2ph4y2cOYpEKYXjGbdFmH4j95t1xtdm6z1pQBmArKJQ9jCYNWTlZBrpQx2GvmhE88Ylwk5GRTRmJ_qo3LBn9oMOG1q33JEMl7SVEEFah0xGHo2rOCOxWNX4A5jwte6qRg2ApGlTayvNcWNw6ly5ydCgYejTsQ9dA_D2aejUsHDodiXJ8Sj-ehsu2VQXras9IGQ0yc","access_token":"","expires_in":1800,"issued_at":"2020-07-12T12:26:11Z"}
```

Able to load UI on https and use default admin/bitnami to sign in but docker login fails

**Describe the results you received:**

**Output from docker login**
```
user@ubuntu:~$ docker login reg.mydomain.com
Username: admin
Password: 
Error response from daemon: login attempt to https://reg.mydomain.com/v2/ failed with status: 401 Unauthorized
```

**Logs**
```
nginx           | 192.168.101.128 - - [12/Jul/2020:11:58:27 +0000] "GET /v2/ HTTP/1.1" 401 76 "-" "docker/19.03.8 go/go1.13.8 git-
commit/afacb8b7f0 kernel/5.4.0-40-generic os/linux arch/amd64 UpstreamClient(Docker-Client/19.03.8 \x5C(linux\x5C))"
nginx           | 192.168.101.128 - admin [12/Jul/2020:11:58:27 +0000] "GET /service/token?account=admin&client_id=docker&offline_token=true&service=harbor-registry HTTP/1.1" 200 1105 "-" "docker/19.03.8 go/go1.13.8 git-commit/afacb8b7f0 kernel/5.4.0-40-generic os/linux arch/amd64 UpstreamClient(Docker-Client/19.03.8 \x5C(linux\x5C))"
registry_1      | time="2020-07-12T11:58:27.771014077Z" level=warning msg="error authorizing context: authorization token required" go.version=go1.14.4 http.request.host=reg.mydomain.com http.request.id=d2452d4f-dbb2-40aa-96cc-c6cf98a9b3eb http.request.method=GET http.request.remoteaddr=192.168.101.128 http.request.uri="/v2/" http.request.useragent="docker/19.03.8 go/go1.13.8 git-commit/afacb8b7f0 kernel/5.4.0-40-generic os/linux arch/amd64 UpstreamClient(Docker-Client/19.03.8 \(linux\))" 
registry_1      | 172.21.0.7 - - [12/Jul/2020:11:58:27 +0000] "GET /v2/ HTTP/1.1" 401 87 "" "docker/19.03.8 go/go1.13.8 git-commit/afacb8b7f0 kernel/5.4.0-40-generic os/linux arch/amd64 UpstreamClient(Docker-Client/19.03.8 \\(linux\\))"
nginx           | 192.168.101.128 - - [12/Jul/2020:11:58:27 +0000] "GET /v2/ HTTP/1.1" 401 87 "-" "docker/19.03.8 go/go1.13.8 git-commit/afacb8b7f0 kernel/5.4.0-40-generic os/linux arch/amd64 UpstreamClient(Docker-Client/19.03.8 \x5C(linux\x5C))"
```

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Version**

- Output of `docker version`:
```
Client:
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.13.8
 Git commit:        afacb8b7f0
 Built:             Tue Jun 23 22:26:12 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.8
  Git commit:       afacb8b7f0
  Built:            Thu Jun 18 08:26:54 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.3-0ubuntu2
  GitCommit:        
 runc:
  Version:          spec: 1.0.1-dev
  GitCommit:        
 docker-init:
  Version:          0.18.0
  GitCommit:        
```

- Output of `docker info`:

```
Client:
 Debug Mode: false

Server:
 Containers: 9
  Running: 9
  Paused: 0
  Stopped: 0
 Images: 9
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 
 runc version: 
 init version: 
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.4.0-40-generic
 Operating System: Ubuntu 20.04 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 34.17GiB
 Name: ubuntu
 ID: QLKH:4RIP:N6NZ:MNAO:KGVU:NQ3V:WDQF:KBCU:PIDY:7PE6:HG47:VWTU
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.26.2, build eefe0d31
docker-py version: 4.2.2
CPython version: 3.7.7
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**

**Comment by skandragon** [comment]
See #7 (my report that seems to mirror yours) and #4 (the one I'm told is the same, but I don't know...)

**Comment by skandragon** [comment]
To be clear, do you have an `nginx` instance in the path prior to the bitnami-provided `nginx`?  I do in my case, but I don't think it's causing the issue.

**Comment by skandragon** [comment]
So far, my debugging of this issue (I'm just trying harbor out as well) seems to be that when harbor-core sends an internal request to the registry on port 5000, it uses an auth of `harbor_registry_user:harbor_registry_password` to authenticate, and that is returning a request for a token handshake, which either fails or isn't started.

I could be wrong, but I think the flow is:

```
client -> nginx -> /v2/ with no auth.  harbor-core handles this, and returns 401 with the token URI.
client -> nginx -> token-uri, with basic auth.  harbor-core handles this, and returns a token.
client -> nginx -> /v2/ with token auth.  harbor-core accepts this, and forwards the request to harbor-registry, using the `harbor_registry_user:harbor_registry_password` credentials.  This is rejected by the registry with a token-uri 401 reply.
```
At this point, the world collapses and we're rejected.

**Comment by skandragon** [comment]
(See a potential temporary work-around on #4)

**Comment by yannhowe** [comment]
> To be clear, do you have an `nginx` instance in the path prior to the bitnami-provided `nginx`? I do in my case, but I don't think it's causing the issue.

My production setup will have it but I made minimal changes for troubleshooting and still ran into the exact same token issues. The token works fine and I did see your issue and I feel it's the same.

**Comment by skandragon** [comment]
I think the issue is that `core` does quite a bit, and the `registry` code is a mostly unmodified 3rd party library.  From what I can tell, it's basically the standard Docker registry.

`core` needs to be in the path (which is why I could get the registry to work with the token being sent to it directly) but that's not how the system works, as that doesn't provide the hooks needed to update redis with blob information as a cache, or postgresql for permanent storage.  Thus, without core, it's just a plain old registry.

I did get `core` to be able to authenticate to the registry, which makes things mostly work.  I added this on #4, but overall it was just changing the auth method in the registry to using a htpasswd file, with the well-known credentials in place.  As there is no direct access to the registry (and the shared user/password can be changed) this is secure enough I suppose.
